aequilibrae/paths/linear_approximation.py

```diff
@@ -257,13 +260,29 @@ def __calculate_step_direction(self):
                     copy_three_dimensions(stp_dir_res.skims.matrix_view, aon_res.skims.matrix_view, self.cores)
                 sd_flows.append(aon_res.total_link_loads)
 
+                if c._selected_links:
+                    aux_res = self.aons[c._id].aux_res
+                    for name, idx in c._aon_results._selected_links.items():
+                        copy_two_dimensions(
+                            aux_res.sl_link_loads_step_direction[idx, :, :],
+                            np.sum(aux_res.temp_sl_link_loading, axis=0)[idx, :, :],
+                            self.cores,
+                        )
+                        copy_three_dimensions(
+                            aux_res.sl_od_matrix_step_direction[idx, :, :, :],
+                            np.sum(aux_res.temp_sl_od_matrix, axis=0)[idx, :, :, :],
+                            self.cores,
+                        )
+
         # 3rd iteration is cfw. also, if we had to reset direction search we need a cfw step before bfw
         elif (self.iter == 3) or (self.do_conjugate_step) or (self.algorithm == "cfw"):
+            print("hit iter 3 step dir block")
             self.do_conjugate_step = False
             self.calculate_conjugate_stepsize()
             for c in self.traffic_classes:
                 sdr = self.step_direction[c._id]
                 pre_previous = self.pre_previous_step_direction[c._id]
+
                 copy_two_dimensions(pre_previous.link_loads, sdr.link_loads, self.cores)
```

Comment:
Additional another bug may have been introduced in this commit 21c66f117fe2c28bedfec00afba54e27c517635a (thanks @janzill ). The current step direction is being copied into the `pre_previous` matrix, this is used to hold the step direction before it is copied back out into the `previous` matrix. However this copy never occurs. I believe it should be copied directly into the `previous` matrix instead. 

If the next iteration is `fw` or `msa` due to restarting, then  `pre_previous`  is not accessed.
If the next iteration is `cfw` then  `pre_previous`  is immediately over written again.
If the next iteration is `bfw` then the  `pre_previous`  is immediately over written.

` pre_previous` is never read from without before being written to.
If a `cfw` then a `bfw` iteration occurs then the `previous` matrix may not have been written to before hand.

Reply:
Yes, I agree. I think this actually is a bug when running BFW because we use previous_step_direction in finding the blend coefficients (betas) in the BFW, which is run before any copying happens. Also, pre_previous is a bit of a misnomer because it is only ever used in BFW to hold a temporary copy during step direction calculations

Reply:
This should be resolved, though I did not change the name